airbyte-integrations/connectors/source-asana/BOOTSTRAP.md

```diff
@@ -0,0 +1,8 @@
+# Asana
+
+Asana is a REST based API. Connector is implemented with [Airbyte CDK](https://docs.airbyte.io/connector-development/cdk-python).
+
+Asana API by default returns 3 fields for each record: `gid`, `name`, `resource_type`.
+If you want to get additional fields you need to specify those fields in each request.
+
+See [this](https://docs.airbyte.io/integrations/sources/asana) link for the nuances about the connector.
```

Comment:
If I get it right, main purpose of the bootstrap is to give bird eye view on the connector and API without diving into technical details, something like this: 
```
Asana is project managing platform designed to organaize and manage teams and their work. 
This connector adds ability to fetch projects, tasks, teams, etc over REST API. See [this](https://docs.airbyte.io/integrations/sources/asana) for details. 
Some streams like  are broke down into workspaces (Teams, Users streams) and projects (Sections, Tasks).

Each record can be uniquely indetified by a "gid" key but also some stream have specific identifier fields (). 
```
Also it would be great to provide stream class structure
<details><summary>It was autogenerated by pyreverse tool so you probably come up with better chart </summary>

```
                                                                ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
                                                                │             SourceAsana             │  │                              check_connection(logger: AirbyteLogger, config: Mapping[str, Any]): Tuple[bool, Any]                              │
                                                                │                                     │  │  streams(config: Mapping[str, Any]): List[Stream]                                                                                              │
                                                                └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

                                                             ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
                                                             │                                                                                                                                                                                                                   │
                                                             │  ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐                      │
                                                             │  │              Projects               │  │                                                                   path(): str                                                                  │                      │
                                                             │  └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                      │
                                                             │    │                                                                                                                                                                                                              │
                                                             │    │                                                                                                                                                                                                              │
                                                             │    ▽                                                                                                                                                                                                              │
┌───────┐────┐────────────────────────────────────────────┐  │  ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐                    ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐────┐────────────────────────────────────────────┐
│ Tags  │    │                 path(): str                │  └▷ │ WorkspaceRequestParamsRelatedStream │  │ request_params(stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any], next_page_token: Mapping[str, Any]): MutableMapping[str, Any] │                    │                                                                     Users                                                                     │    │                 path(): str                │
└───────┘────┘────────────────────────────────────────────┘     └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                    └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘────┘────────────────────────────────────────────┘
                                                                  │
                                                                  │                                                                                                                                                                                          ┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
                                                                  ▽                                                                                                                                                                                          │                                                                                                                                                                                                                      │
┌───────┐────┐────────────────────────────────────────────┐     ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐────┐────────────────────────────────────────────┐  │
│ Teams │    │ path(stream_slice: Mapping[str, Any]): str │  ┌▷ │       WorkspaceRelatedStream        │  │       stream_slices(sync_mode: SyncMode, cursor_field: List[str], stream_state: Mapping[str, Any]): Iterable[Optional[Mapping[str, Any]]]      │  │                 │                                                                 CustomFields                                                                  │    │ path(stream_slice: Mapping[str, Any]): str │  │
└───────┘────┘────────────────────────────────────────────┘  │  └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘────┘────────────────────────────────────────────┘  │
                                                             │    │                                                                                                                                                                                          │                   │                                                                                                                                                                                                  │
                                                             │    │                                         ┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                   └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐                                                 │
                                                             │    ▽                                         ▽                                                                                                                                                                                                                                                                                                                     │                                                 │
                                                             │  ┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐───────────────────┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                                                 │
                                                             │  │                                                                                                                                                                                         │                   │                                                      backoff_time(response): Optional[int]                                                     │  │                                                 │
                                                             │  │                                                                                                                                                                                         │  page_size : int  │  get_opt_fields(): MutableMapping[str, str]                                                                                                    │  │                                                 │
                                                             │  │                                                                                       AsanaStream                                                                                       │  primary_key : str│  next_page_token(response): Optional[Mapping[str, Any]]                                                                                        │  │                                                 │
                                                             │  │                                                                                                                                                                                         │  url_base : str   │  parse_response(response): Iterable[Mapping]                                                                                                   │  │                                                 │
                                                             │  │                                                                                                                                                                                         │                   │  read_slices_from_records(stream_class: Type[AsanaStream], slice_field: str): Iterable[Optional[Mapping[str, Any]]]                            │  │                                                 │
           ┌─────────────────────────────────────────────────┼▷ │                                                                                                                                                                                         │                   │  request_params(stream_state: Mapping[str, Any], stream_slice: Mapping[str, any], next_page_token: Mapping[str, Any]): MutableMapping[str, Any]│  │                                                 │
           │                                                 │  └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘───────────────────┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                                                 │
           │                                                 │    △                                         △                                                                                                                                                                                                                                                                                                                     │                                                 │
           │                                                 │    │                                         └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐                                                                                                                                                                    │                                                 │
           │                                                 │    │                                                                                                                                                                                          │                                                                                                                                                                    │                                                 │
           │                                                 │  ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                                                                                                                                                                    │                                                 │
           │                                                 │  │             Workspaces              │  │                                                                   path(): str                                                                  │  │                                                                                                                                                                    │                                                 │
           │                                                 │  └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                                                                                                                                                                    │                                                 │
           │                                                 │                                                                                                                                                                                               │                                                                                                                                                                    │                                                 │
           │                                                 └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                                                 │
           │                                                                                                                                                                                                                                                 │                                                                                                                                                                                                                      │
           │                                                    ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                                                                                                                                                                                                                      │
           │                                                    │              Sections               │  │                                                   path(stream_slice: Mapping[str, Any]): str                                                   │  │                                                                                                                                                                                                                      │
           │                                                    └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                                                                                                                                                                                                                      │
           │                                                      │                                                                                                                                                                                          │                                                                                                                                                                                                                      │
           │                                                      │                                                                                                                                                                                          │                                                                                                                                                                                                                      │
           │                                                      ▽                                                                                                                                                                                          │                                                                                                                                                                                                                      │
           │                                                    ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                                                                                                                                                                                                                      │
           └─────────────────────────────────────────────────── │        ProjectRelatedStream         │  │       stream_slices(sync_mode: SyncMode, cursor_field: List[str], stream_state: Mapping[str, Any]): Iterable[Optional[Mapping[str, Any]]]      │  │                                                                                                                                                                                                                      │
                                                                └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                                                                                                                                                                                                                      │
                                                                  △                                                                                                                                                                                          │                                                                                                                                                                                                                      │
                                                                  │                                                                                                                                                                                          │                                                                                                                                                                                                                      │
                                                                  │                                                                                                                                                                                          │                                                                                                                                                                                                                      │
                                                                ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │                                                                                                                                                                                                                      │
                                                                │                Tasks                │  │                                                                   path(): str                                                                  │  │                                                                                                                                                                                                                      │
                                                                │                                     │  │  request_params(stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any], next_page_token: Mapping[str, Any]): MutableMapping[str, Any]│  │                                                                                                                                                                                                                      │
                                                                └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │                                                                                                                                                                                                                      │
                                                                                                                                                                                                                                                             │                                                                                                                                                                                                                      │
                                                                  ┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
                                                                  │                                                                                                                                                                                          │
                                                                ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐  │
                                                                │               Stories               │  │                                                   path(stream_slice: Mapping[str, Any]): str                                                   │  │
                                                                │                                     │  │  stream_slices(sync_mode: SyncMode, cursor_field: List[str], stream_state: Mapping[str, Any]): Iterable[Optional[Mapping[str, Any]]]           │  │
                                                                └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘  │
                                                                                                                                                                                                                                                             │
                                                                  ┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
                                                                  │
                                                                ┌─────────────────────────────────────┐──┐────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
                                                                │           TeamMemberships           │  │                                                   path(stream_slice: Mapping[str, Any]): str                                                   │
                                                                │                                     │  │  stream_slices(sync_mode: SyncMode, cursor_field: List[str], stream_state: Mapping[str, Any]): Iterable[Optional[Mapping[str, Any]]]           │
                                                                └─────────────────────────────────────┘──┘────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```
</details>

Or you can generate simple class hierarchy using this tool https://codepen.io/weizhenye/details/eoYvye
like this:
```
AsanaStream
├── WorkspaceRelatedStream
│   └── CustomFields
└── ProjectRelatedStream
```
@Phlair Could you  share your thoughts on this?

Reply:
Here is autogenerated classes structure but I dont think its easy to read:
![image](https://user-images.githubusercontent.com/437607/135402813-0807196f-51eb-4aee-bb7f-220b8a8c2441.png)

@Zirochkaa  @Phlair  What do you think on this?

Reply:
I don't see a point in including stream class structure. If a developer really really needs it they can generate one.

Reply:
<img width="776" alt="image" src="https://user-images.githubusercontent.com/19872253/135403680-8ec7da0f-ffd1-4bf3-9e26-c155aaa7641f.png">

Including the whole stream class structure won't help in achieving above point. Link to the document is [here](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit).

Reply:
> Or you can generate simple class hierarchy using this tool https://codepen.io/weizhenye/details/eoYvye

It actually can be helpful.

Reply:
> Here is autogenerated classes structure but I dont think its easy to read: ![image](https://user-images.githubusercontent.com/437607/135402813-0807196f-51eb-4aee-bb7f-220b8a8c2441.png)
> 
> @Zirochkaa @Phlair What do you think on this?

Yes, it's hard to read. I liked the idea about simple class hierarchy :)

Reply:
I think both of you put forward valid points that ideally we always include in connector PRs.
- The bootstrap should be (imo) a kick start to understanding how the API works, not how the code we've written works. So the high level view that @avida suggests is very useful, but also diving into the technical nuances of the API (not the nuances of how we built around it) is useful too. For example, the API may have more capabilities that we haven't implemented (yet). Knowledge of the API as a whole including these fits in the bootstrap I think.
- I find the class hierarchy very useful (also advocate for the simplest representation) but it probably doesn't fit in the bootstrap, rather within the code itself. This is where we're explaining specifically our approach to interfacing with the API.

Reply:
@Phlair got it. Here Ive created seperate ticket for class hierarchy generation tool https://github.com/airbytehq/airbyte/issues/6561
@Zirochkaa please just add more details on how asana API works beside it based on REST.

Reply:
> @Zirochkaa please just add more details on how asana API works beside it based on REST.

there is nothing more to add. It’s just simple REST api, that’s it. More detail you can find following links in the bootstrap file.